File: api/src/test/java/net/blueshell/api/factory/model/QuestionFactory.java
package net.blueshell.api.factory.model;

import net.blueshell.api.testutil.ModelTestUtils;
import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.common.enums.QuestionType;
import net.blueshell.api.model.survey.Question;
import net.blueshell.api.model.survey.Survey;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for Question model test instances.
 */
@Component
@RequiredArgsConstructor
public class QuestionFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;

    public Question createBasic() {
        Question question = new Question();
        ModelTestUtils.setId(question, generateId());
        question.setIdx((long) faker.number().numberBetween(1, 100));
        question.setType(faker.options().option(QuestionType.class));
        question.setLabel(faker.lorem().sentence() + "?");
        question.setAnswerCount(0L);
        return question;
    }

    public Question createFull() {
        Question question = createBasic();
        if (question.getType() == QuestionType.CHECKBOX || question.getType() == QuestionType.RADIO) {
            question.setChoiceLabels(generateChoiceLabels());
        }
        return question;
    }

    public Question createWithCustomizations(java.util.function.Consumer<Question> customizer) {
        Question question = createFull();
        customizer.accept(question);
        return question;
    }

    public Question createForSurvey(Survey survey) {
        return createWithCustomizations(question -> {
            question.setSurvey(survey);
            question.setSurveyId(survey.getId());
        });
    }

    public Question createMultipleChoice() {
        return createWithCustomizations(question -> {
            question.setType(QuestionType.RADIO);
            question.setChoiceLabels(Arrays.asList("Option A", "Option B", "Option C", "Option D"));
        });
    }

    public Question createText() {
        return createWithCustomizations(question -> {
            question.setType(QuestionType.OPEN);
            question.setChoiceLabels(null);
        });
    }

    private List<String> generateChoiceLabels() {
        int count = faker.number().numberBetween(2, 6);
        return Arrays.asList(
                "Strongly Disagree", "Disagree", "Neutral", "Agree", "Strongly Agree"
        ).subList(0, count);
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/survey/QuestionFactory.kt
package net.blueshell.api.factory.model.survey

import com.github.javafaker.Faker
import net.blueshell.api.shared.enums.QuestionType
import net.blueshell.api.survey.model.Question
import net.blueshell.api.survey.model.Survey
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for Question model test instances.
 */
@Component
class QuestionFactory(
    private val faker: Faker
) {

    fun createBasic(): Question {
        val question = Question()
        question.idx = faker.number().numberBetween(1, 100).toLong()
        question.type = faker.options().option(QuestionType::class.java)
        question.label = faker.lorem().sentence() + "?"
        question.answerCount = 0L
        return question
    }

    fun createFull(): Question {
        val question = createBasic()
        if (question.type == QuestionType.CHECKBOX || question.type == QuestionType.RADIO) {
            question.choiceLabels = generateChoiceLabels().toMutableList()
        }
        return question
    }

    fun createWithCustomizations(customizer: Consumer<Question>): Question {
        val question = createFull()
        customizer.accept(question)
        return question
    }

    fun createForSurvey(survey: Survey): Question {
        return createWithCustomizations { question ->
            question.survey = survey
        }
    }

    fun createMultipleChoice(): Question {
        return createWithCustomizations { question ->
            question.type = QuestionType.RADIO
            question.choiceLabels = mutableListOf("Option A", "Option B", "Option C", "Option D")
        }
    }

    fun createText(): Question {
        return createWithCustomizations { question ->
            question.type = QuestionType.OPEN
            question.choiceLabels = null
        }
    }

    private fun generateChoiceLabels(): List<String> {
        val count = faker.number().numberBetween(2, 6)
        val labels = listOf("Strongly Disagree", "Disagree", "Neutral", "Agree", "Strongly Agree")
        return labels.subList(0, count)
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
